fiftyone/utils/ultralytics.py

```diff
@@ -270,3 +289,105 @@ def _to_detections(self, df):
                 for row in df.itertuples()
             ]
         )
+
+
+class FiftyOneYOLOModel(Model):
+    """YOLO model wrapper for FiftyOne.
+
+    Args:
+        config: a dict which has the `ultralytics.YOLO` model as the value for
+            the key `model`
+    """
+
+    def __init__(self, config):
+        self.model = config.get("model", None)
+
+    @property
+    def media_type(self):
+        return "image"
+
+    @property
+    def ragged_batches(self):
+        return True
+
+    @property
+    def transforms(self):
+        return None
+
+    @property
+    def preprocess(self):
+        return False
+
+    def predict(self, args):
+        raise NotImplementedError("Subclass must implement `predict`")
+
+    def predict_all(self, args):
+        return self.predict(args)
```

Comment:
I don't think this actually works as written, as `predict()` calls `Image.fromarray()`which  cannot handle a list of images or `NHWC` tensor.

Since you set `ragged_batches == True` in the FO model definition, I believe `apply_model(..., batch_size>1)` will raise an error complaining that the model does not support batch inference. So it may be fine to just delete this implementation of `predict_all()` and let the default implementation from `fom.Model` be used.

Reply:
One of the reasons I didn't initially write these wrapper classes is that `apply_model()` doesn't currently have a way to know that a model can directly accept filepaths for inference, like Ultralytics model objects support. It currently always loads the images into numpy arrays, possibly using a data loader if it is a Torch model.

In the Ultralytics integration docs, I showed how batch inference can be done like so:
https://docs.voxel51.com/integrations/ultralytics.html#batch-inference

I did not investigate the Ultralytics implementation to see what is happening inside: does it generate a 4D tensor and do a single forward pass (presumably more efficient) or does it just process each image serially (presumably less efficient)? And is it more efficient to pass filepaths directly to the Ultralytics model rather than loading them yourself?

Reply:
I wasn't sure if numpy arrays could be passed to Ultralytics models (how would it know if any necessary preprocessing had already been performed?) or if their API *required* filepaths to be provided in general

Reply:
Did a quick timing test for YOLO models to see how filepath vs PIL Image and batching impact efficiency.

```
import fiftyone as fo
import fiftyone.zoo as foz
import fiftyone.brain as fob
from fiftyone import ViewField as F

import timeit
from PIL import Image
from ultralytics import YOLO

dataset = foz.load_zoo_dataset("quickstart")
yolov5n = YOLO('yolov5n.pt')

def run_yolov5n_filepath(filepath):
    yolov5n(filepath, verbose=False);

print(timeit.timeit(lambda: run_yolov5n_filepath(dataset.first().filepath), number=100))
## 6.987219458999988

def run_yolov5_image(filepath):
    yolov5n(Image.open(filepath), verbose=False);

print(timeit.timeit(lambda: run_yolov5_image(dataset.first().filepath), number=100))
## 7.138490875000002

def run_yolov5_images(filepaths):
    ims = [Image.open(filepath) for filepath in filepaths]
    yolov5n(ims, verbose=False);

print(timeit.timeit(lambda: run_yolov5_images([dataset.first().filepath]*10), number=10))
## 4.4615429579999955

def run_yolov5_filepaths(filepaths):
    yolov5n(filepaths, verbose=False);

print(timeit.timeit(lambda: run_yolov5_filepaths([dataset.first().filepath]*10), number=10))
## 4.528784499999972

def run_yolov5_numpy(filepaths):
    ims = [np.array(Image.open(filepath)) for filepath in filepaths]
    yolov5n(ims, verbose=False);

print(timeit.timeit(lambda: run_yolov5_numpy([dataset.first().filepath]*10), number=10))
4.5486639160000095
```

Obviously this is just one model size and one image, but roughly speaking it appears that passing in filepaths vs PIL Images vs numpy arrays is rather inconsequential, while utilizing the model's internal batching is important, if/when it is possible.

Reply:
nice! so this should mean that you can just return `ragged_batches == False` in the model's definition and then `apply_model()` will allow a `batch_size > 1` to be provided

Reply:
Okay, `ragged_batches` is now `False`, and running `apply_model()` works. Is this what you had in mind, or do I need to define the `predict_all()` method as well to get the batching speedups?

Reply:
I'm confused how `predict_all()` could work right now: it just forwards the list of numpy arrays to `predict()`, but does `Image.fromarray()` actually support a list of arrays as input?

Reply:
Fixed! Now batching happens properly for `FiftyOneYOLODetectionModel` and `FiftyOneYOLOPoseModel`, while batching is skipped for segmentation models because masks get padded.